### ISSUE TITLE
semgrep-core returns mistaken match bounds when unicode strings are present in Python

### ISSUE THREAD

**Issue by rtpg** [semgrep/semgrep#2477]
**Describe the bug**
Autofixing seems to hit some weird issues when there are unicode strings in Python.

**To Reproduce**
https://semgrep.dev/s/oxYN

**Expected behavior**
 `    self.assertEqual(entry.a, 'あいうえお')` turns to `assert entry.a == 'あいうえお'` when applying an autofix like 
```
self.assertEqual($X, $Y)  -> assert $X == $Y
```

**Actual**
```
def test(self, entry):
    self.assertEqual(entry.a, 'あいうえお')
    print("Done")  # removing this line "fixes" the issue
```

applying the autofix ends up with 
```
def test(self, entry):
    assert entry.a == 'あいうえお')
    prin
    print("Done")  # removing this line "fixes" the issue
```
(making the string be just ASCII characters makes the issue not appear, and the `print` statement not being there also fixes the issue)

This feels like something where there's some string slicing that isn't handling unicode properly (difference between glyphs and bytes probably?)

**Comment by ievans** [comment]
Thanks @rtpg ! I can confirm this is a bug, and I like your hypothesis about the glyph/byte distinction being at the root of it. @DrewDennison or @mschwager, want to take a look?

Autofix is still an experimental feature but we are excited to improve it. If you don't mind, can you share the relative priority of the issue for you? is it a blocker to using Semgrep further (P0), a significant annoyance (P1), or just an FYI (P2)?

**Comment by rtpg** [comment]
I understand autofix is experimental, but from our perspective it's hard to set up semgrep on an existing codebase without introducing a lot of busywork. This is a P0 for our own personal adoption, yeah.

**Comment by rtpg** [comment]
OK an update on this one, turns out I was able to isolate this to `semgrep-core`, so it's not an autofix-only issue.

tl;dr is that semgrep-core reports column counts as.... I guess bytes? I dunno, couldn't find the ocaml source, but it's saying that the end of the line is at column 49 when there are only 42 columns on the error'd line. this causes replacement to fail downstream, but I imagine this also leaks to other stuff.

If I had to guess (esp the "abstract content" turning into ZZZZs), I'm going to bet that deep in your ocaml code you open files with ascii encoding with some replace param, instead of doing some encoding detection or opening as utf-8, and that causes these issues.  If you're not going to try and guess the encoding, utf-8 would probably be a better "random encoding" to chose.

example.py
```
def test(self, entry):
    self.assertEqual(entry.a, "あいうえお")
    print("Done")  # removing this line "fixes" the issue
```

target.yml
```
example.py
```

pattern.yml
```
rules:
- id: 0.simple-assert
  pattern: self.assertEqual($X, $Y)
  severity: WARNING
  languages:
  - python
  message: <internalonly>
```

Running the following:
```
semgrep-core -lang python -rules_file patterm.yml -taget_file target.yml
```

we get 
```
{
  "matches": [
    {
      "check_id": "0.simple-assert",
      "path": "example.py",
      "start": {
        "line": 2,
        "col": 5,
        "offset": 27
      },
      "end": {
        "line": 2,
        "col": 49,
        "offset": 71
      },
      "extra": {
        "message": "<internalonly>",
        "metavars": {
          "$Y": {
            "start": {
              "line": 2,
              "col": 31,
              "offset": 53
            },
            "end": {
              "line": 2,
              "col": 48,
              "offset": 70
            },
            "abstract_content": "\"ZZZZZZZZZZZZZZZ\"",
            "unique_id": {
              "type": "AST",
              "md5sum": "7d81525898bc9f3c4c76f16da8fc1aa3"
            }
          },
          "$X": {
            "start": {
              "line": 2,
              "col": 22,
              "offset": 44
            },
            "end": {
              "line": 2,
              "col": 29,
              "offset": 51
            },
            "abstract_content": "entry.a",
            "unique_id": {
              "type": "AST",
              "md5sum": "504dab47a61c4083b014dfead48de916"
            }
          }
        }
      }
    }
  ],
  "errors": [],
  "stats": {
    "okfiles": 1,
    "errorfiles": 0
  }
}
```


the "end" column in particular for the error is totally off, saying that the error ends on column 49 (the line only goes to 42)

**Comment by rtpg** [comment]
reading more about this, sounds like ocaml's unicode story isn't great but [Camomile](https://github.com/yoriyuki/Camomile) can provide a unicode text type for you to use while reading files.

**Comment by ievans** [comment]
paging @aryx @mjambon @emjin for ocaml expertise :)

**Comment by aryx** [comment]
yes we're doing some hacks in pfff around unicode in our lexers. One option is to switch to tree-sitter for parsing Python; should be quicky, the language is quite small and we already have an AST defined for it.

**Comment by malexmave** [comment]
This is probably also the root cause of [this issue I mistakenly opened in semgrep-rules (but which seems to be a semgrep bug, not an issue with the rule)](https://github.com/returntocorp/semgrep-rules/issues/1398), where unicode seems to be completely ignored in rule matching on Python files.

**Comment by IagoAbal** [comment]
This is an example by @malexmave: https://semgrep.dev/s/y6QA

**Comment by nmote** [comment]
The original playground example works now. I'm guessing that this has been addressed, but feel free to open a new issue with an updated example if not.
